dds_web/security/tokens.py

```diff
@@ -27,18 +31,21 @@ def encrypted_jwt_token(
         Can be set to None for protecting the signed token.
     :param timedelta expires_in: This is the maximum allowed age of the token. (default 2 days)
     :param Dict or None additional_claims: Any additional token claims can be added. e.g., {"iss": "DDS"}
+    :param Boolean fully_authenticated: set to True only after successful 2fa which means that all authentication
+        steps have succeeded and this final token can be used for normal operation by the cli (default False)
     """
+    jwe_protected_header = {
+        "alg": "A256KW",
+        "enc": "A256GCM",
+    }
+    if fully_authenticated:
+        # exp claim in this (integrity) protected JWE header is provided only to let the
+        # cli know the precise expiration time of the encrypted token. It has no impact
+        # on the actual enforcement of the expiration of the token.
+        jwe_protected_header["exp"] = (dds_web.utils.current_time() + expires_in).timestamp()
```

Comment:
Given that you have come up with this really clever way of balancing token length vs. additional metadata for a better user experience: Why are you still so restrictive with regard to putting anything else into the header than the expiration time?

Reply:
Because it is plaintext and putting more information makes the token longer. We have to carefully think what we really need and if we need to put it in the header or even in the encrypted part of the token.

Reply:
I acknowledge the problem of token length, but think there is a difference with regard to tokens that are part of an URL and send out via e-mail and the "regular" MFA tokens that we cache in the CLI. For the latter, I don't see why we have to be _that_ strict in terms of length that we even need to scrap the username, but you have the say in the security aspects.

Reply:
We don't want to reveal the usernames unless we really have to.

Reply:
In which regard do we reveal the username here? In order to authenticate, the user needs to provide it first, no? The `.dds_cli_token` file should only be readable by the user, Johannes specifically wrote a check to set the correct file permissions. 

Since the username is obfuscated in the token and needs to be deserialized with `jwt.JWT(jwt=token)` and `token.token.jose_header` first, I would consider this even slightly better than if we store the entered username locally as you suggested for Anands PR.

Reply:
what do you mean by **obfuscated**? 
Deserialize is like you copy-paste the token string at jwt.io and see the plaintext header. It is nothing fancy.

Reply:
@MatthiasZepper https://github.com/ScilifelabDataCentre/dds_cli/pull/244#issuecomment-1047054729

Reply:
I know that it is not encrypted, but it is also not clear text and will therefore not show up when e.g. searching with `grep `for a particular username on a multi-user system.

I would agree that e.g. on a university cluster with hundreds of users, finding readable token file for a particular user owing to wrongly set file system permissions like this would be a security nightmare.

But I agree with you and Ina that we must not have this discussion now. It is a waste of precious time for you guys. If the two of you prefer to have a separate username retrieval endpoint in the API to save <210 Bytes in token length, I will accept that I was outvoted.